File: third-person/missiles.go
package main

// Missile struct
type Missile struct {
	World               *World
	UID                 string
	NID                 string
	X, Y, Z             float32
	DX, DY, DZ          float32
	MinBX, MinBY, MinBZ int
	MaxBX, MaxBY, MaxBZ int
	Radius              float32
	Height              float32
	GC                  bool
	Damage              int
	HitFunc             func(me *Missile, thing ThingInterface)
}

// BlockBorders func
func (me *Missile) BlockBorders() {
	me.MinBX = int((me.X - me.Radius) * InverseBlockSize)
	me.MinBY = int(me.Y * InverseBlockSize)
	me.MinBZ = int((me.Z - me.Radius) * InverseBlockSize)
	me.MaxBX = int((me.X + me.Radius) * InverseBlockSize)
	me.MaxBY = int((me.Y + me.Height) * InverseBlockSize)
	me.MaxBZ = int((me.Z + me.Radius) * InverseBlockSize)
}

// AddToBlocks func
func (me *Missile) AddToBlocks() {
	for gx := me.MinBX; gx <= me.MaxBX; gx++ {
		for gy := me.MinBY; gy <= me.MaxBY; gy++ {
			for gz := me.MinBZ; gz <= me.MaxBZ; gz++ {
				me.World.GetBlock(gx, gy, gz).AddMissile(me)
			}
		}
	}
}

// RemoveFromBlocks func
func (me *Missile) RemoveFromBlocks() {
	for gx := me.MinBX; gx <= me.MaxBX; gx++ {
		for gy := me.MinBY; gy <= me.MaxBY; gy++ {
			for gz := me.MinBZ; gz <= me.MaxBZ; gz++ {
				me.World.GetBlock(gx, gy, gz).RemoveMissile(me)
			}
		}
	}
}

// Overlap func
func (me *Missile) Overlap(b *Thing) bool {
	square := me.Radius + b.Radius
	return Abs(me.X-b.X) <= square && Abs(me.Z-b.Z) <= square
}

// Collision func
func (me *Missile) Collision() bool {
	searched := make(map[*Thing]bool)
	for gx := me.MinBX; gx <= me.MaxBX; gx++ {
		for gy := me.MinBY; gy <= me.MaxBY; gy++ {
			for gz := me.MinBZ; gz <= me.MaxBZ; gz++ {
				block := me.World.GetBlock(gx, gy, gz)
				for t := 0; t < block.ThingCount; t++ {
					thing := block.Things[t]
					if _, ok := searched[thing]; !ok {
						searched[thing] = true
						if me.Overlap(thing) {
							me.HitFunc(me, thing)
							return true
						}
					}
				}
			}
		}
	}
	minGX := int((me.X - me.Radius))
	minGY := int(me.Y)
	minGZ := int((me.Z - me.Radius))
	maxGX := int((me.X + me.Radius))
	maxGY := int((me.Y + me.Height))
	maxGZ := int((me.Z + me.Radius))
	for gx := minGX; gx <= maxGX; gx++ {
		for gy := minGY; gy <= maxGY; gy++ {
			for gz := minGZ; gz <= maxGZ; gz++ {
				bx := int(float32(gx) * InverseBlockSize)
				by := int(float32(gy) * InverseBlockSize)
				bz := int(float32(gz) * InverseBlockSize)
				tx := gx - bx*BlockSize
				ty := gy - by*BlockSize
				tz := gz - bz*BlockSize
				tile := me.World.GetTileType(bx, by, bz, tx, ty, tz)
				if TileClosed[tile] {
					me.HitFunc(me, nil)
					return true
				}
			}
		}
	}
	return false
}

// Update func
func (me *Missile) Update() {
	if me.Collision() {
		return
	}
	me.X += me.DX
	me.Y += me.DY
	me.Z += me.DZ
	me.Collision()
}

// NewPlasma func
func NewPlasma(world *World, damage int, x, y, z, dx, dy, dz float32) *Missile {
	me := &Missile{}
	me.UID = "plasma"
	me.NID = NextNID()
	me.World = world
	me.X = x
	me.Y = y
	me.Z = z
	me.DX = dx
	me.DY = dy
	me.DZ = dz
	me.Radius = 0.4
	me.Height = 1.0
	me.Damage = damage
	me.HitFunc = PlasmaHit
	world.AddMissile(me)
	me.BlockBorders()
	me.AddToBlocks()
	return me
}

// PlasmaHit func
func PlasmaHit(me *Missile, thing ThingInterface) {
	me.X -= me.DX
	me.Y -= me.DY
	me.Z -= me.DZ
	if thing != nil {
		thing.Damage(1)
	}
	// client impact sound
	NewPlasmaExplosion(me.World, me.X, me.Y, me.Z)
	me.GC = true
}

File: public/missiles.js
class Missile {
    constructor() {
        this.World = null
        this.UID = ""
        this.SID = ""
        this.NID = ""
        this.Sprite = null
        this.X = 0
        this.Y = 0
        this.Z = 0
        this.DX = 0
        this.DY = 0
        this.DZ = 0
        this.MinBX = 0
        this.MinBY = 0
        this.MinBZ = 0
        this.MaxBX = 0
        this.MaxBY = 0
        this.MaxBZ = 0
        this.Radius = 0
        this.Height = 0
    }
    BlockBorders() {
        this.MinBX = Math.floor((this.X - this.Radius) * InverseBlockSize)
        this.MinBY = Math.floor(this.Y * InverseBlockSize)
        this.MinBZ = Math.floor((this.Z - this.Radius) * InverseBlockSize)
        this.MaxBX = Math.floor((this.X + this.Radius) * InverseBlockSize)
        this.MaxBY = Math.floor((this.Y + this.Height) * InverseBlockSize)
        this.MaxBZ = Math.floor((this.Z + this.Radius) * InverseBlockSize)
    }
    AddToBlocks() {
        for (let gx = this.MinBX; gx <= this.MaxBX; gx++) {
            for (let gy = this.MinBY; gy <= this.MaxBY; gy++) {
                for (let gz = this.MinBZ; gz <= this.MaxBZ; gz++) {
                    this.World.GetBlock(gx, gy, gz).AddMissile(this)
                }
            }
        }
    }
    RemoveFromBlocks() {
        for (let gx = this.MinBX; gx <= this.MaxBX; gx++) {
            for (let gy = this.MinBY; gy <= this.MaxBY; gy++) {
                for (let gz = this.MinBZ; gz <= this.MaxBZ; gz++) {
                    this.World.GetBlock(gx, gy, gz).RemoveMissile(this)
                }
            }
        }
    }
    Overlap(b) {
        let square = this.Radius + b.Radius
        return Math.abs(this.X - b.X) <= square && Math.abs(this.Z - b.Z) <= square
    }
    Collision() {
        let searched = new Set()
        for (let gx = this.MinBX; gx <= this.MaxBX; gx++) {
            for (let gy = this.MinBY; gy <= this.MaxBY; gy++) {
                for (let gz = this.MinBZ; gz <= this.MaxBZ; gz++) {
                    let block = this.World.GetBlock(gx, gy, gz)
                    for (let t = 0; t < block.thingCount; t++) {
                        let thing = block.things[t]
                        if (thing.Health > 0) {
                            if (searched.has(thing)) continue
                            searched.add(thing)
                            if (this.Overlap(thing)) {
                                this.Hit(thing)
                                return true
                            }
                        }
                    }
                }
            }
        }
        let minGX = Math.floor(this.X - this.Radius)
        let minGY = Math.floor(this.Y)
        let minGZ = Math.floor(this.Z - this.Radius)
        let maxGX = Math.floor(this.X + this.Radius)
        let maxGY = Math.floor(this.Y + this.Height)
        let maxGZ = Math.floor(this.Z + this.Radius)
        for (let gx = minGX; gx <= maxGX; gx++) {
            for (let gy = minGY; gy <= maxGY; gy++) {
                for (let gz = minGZ; gz <= maxGZ; gz++) {
                    let bx = Math.floor(gx * InverseBlockSize)
                    let by = Math.floor(gy * InverseBlockSize)
                    let bz = Math.floor(gz * InverseBlockSize)
                    let tx = gx - bx * BlockSize
                    let ty = gy - by * BlockSize
                    let tz = gz - bz * BlockSize
                    let tile = this.World.GetTileType(bx, by, bz, tx, ty, tz)
                    if (TileClosed[tile]) {
                        this.Hit(null)
                        return true
                    }
                }
            }
        }
        return false
    }
    Update() {
        if (this.Collision()) {
            return true
        }
        this.X += this.DX
        this.Y += this.DY
        this.Z += this.DZ
        return this.Collision()
    }
    Render(interpolation, spriteBuffer, camX, camZ, camAngle) {
        let sin = camX - this.X
        let cos = camZ - this.Z
        let length = Math.sqrt(sin * sin + cos * cos)
        sin /= length
        cos /= length
        Render3.Sprite(spriteBuffer[this.SID], this.X, this.Y, this.Z, sin, cos, this.Sprite)
    }
}

class Plasma extends Missile {
    constructor(world, nid, damage, x, y, z, dx, dy, dz) {
        super()
        this.UID = "plasma"
        this.SID = "missiles"
        this.NID = nid
        this.Sprite = SpriteData[this.SID]["baron-missile-front-1"]
        this.World = world
        this.X = x
        this.Y = y
        this.Z = z
        this.DX = dx
        this.DY = dy
        this.DZ = dz
        this.Radius = 0.4
        this.Height = 1.0
        this.Damage = damage
        this.Hit = this.PlasmaHit
        world.AddMissile(this)
        this.BlockBorders()
        this.AddToBlocks()
        return this
    }
    PlasmaHit(thing) {
        this.X -= this.DX
        this.Y -= this.DY
        this.Z -= this.DZ
        if (thing !== null)
            thing.Damage(1)
        Sounds["plasma-impact"].play()
        new PlasmaExplosion(this.World, this.X, this.Y, this.Z)
        this.RemoveFromBlocks()
    }
}
